### DISCUSSION TITLE
YOLOv8-seg OpenVINO: first predict() takes 10–20s, subsequent calls are fast is this expected behavior?

### DISCUSSION THREAD

**Discussion by supergamefury** [ultralytics/ultralytics#23019] [Q&A] [answered]
Hello
I’m running YOLOv8 segmentation exported to OpenVINO to get the counts of the pastries/slices from an image. The model loads quickly, but the first predict() call is extremely slow (order of 9–20 seconds). After that, subsequent calls are fast (~0.6s). This is a problem for my use case because the backend may switch models dynamically per SKU, so the “first inference penalty” can hit repeatedly.

Environment

OS: Ubuntu 22 (NUC)

CPU/GPU: Intel CPU (i3-1315U)+ Intel iGPU (OpenVINO GPU plugin available)

Python: 3.10.19

Ultralytics: 8.3.240

OpenVINO: 2025.4.1

Model: YOLOv8-seg exported to OpenVINO bundle (*_openvino_model/ with best.xml + metadata.yaml)

Input: camera frames ~640, but model input is 1024 (so preprocessing resizes/letterboxes to 1024)

What I’m seeing

Model “load” is ~1s, but

First inference: ~9–20s

Next inference: ~0.6s

Example logs:

first call: predict ~ 9465 ms, then post steps (boxes/masks/polygons) are much smaller

second call: total end-to-end ~650 ms

**Relevant code:**

Device mapping used for Ultralytics/OpenVINO backend:

I map OpenVINO “GPU” to Ultralytics/OpenVINO backend string "intel:gpu": 

def _ultra_device(device_name: str) -> str:
    if device_name and "GPU" in device_name.upper():
        return "intel:gpu"
    return "cpu"

Loading v8-seg OpenVINO bundle via Ultralytics
I detect seg support across Ultralytics versions, load with task="segment", inject names if needed, and store device/imgsz on the model object. 

'''seg_supported check...'''
yolo_model = YOLO(bundle_dir, task="segment")  # fallback YOLO(bundle_dir) if needed

'''inject names from metadata if ultralytics didn't load it'''
if getattr(yolo_model, "names", None) is None:
    yolo_model.names = _read_metadata_names(bundle_dir)

yolo_model._uy_device = _ultra_device(device_name)
yolo_model._uy_imgsz = _ir_input_hw_from_xml(bundle_xml)[0]

# What I tried:

retina_masks=False (I’m already using that)
Image input is 640 but model is 1024 → preprocessing resizes/letterboxes (I know this adds some cost, but the big hit is the first predict)

Warmup: calling one dummy predict() at model load time shifts the latency from “first scan” to “load”, but with dynamic model switching this is not as helpful.

# **Question**
Is this big first-call latency expected due to OpenVINO compilation / plugin init / caching happening lazily on first predict()?

Is there a recommended way to force compilation / init at load time (or “cheap warmup”) without doing full postprocessing?

# **Why I moved to Ultralytics:**
I originally had a pure-OpenVINO pipeline where I run the IR with ov.Core().compile_model(...) and then do my own post-processing. But my custom YOLOv8-seg postprocess wasn’t aligning with Ultralytics .pt output—mask decode / counts did not match reliably. So I switched to running the OpenVINO-exported bundle via Ultralytics YOLO(bundle_dir, task="segment"), and that fixed the output alignment. The downside is the big first-inference latency.

my original post process:

def postprocess_V8_seg(
    pred_boxes: np.ndarray,
    protos: np.ndarray,
    input_hw: Tuple[int, int],
    orig_img: np.ndarray,
    min_conf_threshold: float = 0.25,
    nms_iou_threshold: float = 0.7,
    max_detections: int = 1000,
) -> tuple:
    """
    Post-process YOLOv8 segmentation model outputs exported to OpenVINO.
    pred_boxes: (1, 37, N) or (1, N, 37)
        Layout per prediction (single-class seg):
            [cx, cy, w, h, conf_or_logit, m0, m1, ..., m31]   (len=37)
    protos: (1, 32, Hm, Wm)
    input_hw: (H_in, W_in) of network input (e.g. (640, 640))
    orig_img: original RGB image (H, W, 3)
    Returns:
        tuple: (detections, masks)
        - detections: List[tuple] in format (class_id, x, y, width, height, confidence)
        - masks: List[np.ndarray] or List[None] - segmentation masks for each detection
        Matching postprocess_V8 format for compatibility with infer2 workflow.
    """

    # Ensure shape is (1, N, 37)
    if pred_boxes.ndim != 3:
        raise ValueError(f"Expected pred_boxes with 3 dims, got {pred_boxes.ndim}")
    if pred_boxes.shape[1] == 37:
        pred_boxes = np.transpose(pred_boxes, (0, 2, 1))

    b, num_boxes, num_feats = pred_boxes.shape
    if b != 1:
        raise ValueError(f"Only batch size 1 supported, got {b}")
    if num_feats != 37:
        raise ValueError(f"Expected 37 features, got {num_feats}")

    # Convert to torch for convenience
    pred = torch.from_numpy(pred_boxes[0])  # (N, 37)
    xywh = pred[:, :4]
    raw_conf = pred[:, 4]
    mask_coeffs = pred[:, 5:]  # (N, 32)
    nm = mask_coeffs.shape[1]

    # Decide whether raw_conf is already [0,1] or a logit
    with torch.no_grad():
        min_v = float(raw_conf.min())
        max_v = float(raw_conf.max())

    if 0.0 <= min_v and max_v <= 1.0:
        conf = raw_conf  # already probability
    else:
        conf = raw_conf.sigmoid()  # treat as logit

    # Confidence filter
    keep_conf = conf >= min_conf_threshold
    if not keep_conf.any():
        print("DEBUG(post_seg): no boxes above confidence threshold")
        return [], []

    xywh = xywh[keep_conf]
    conf = conf[keep_conf]
    mask_coeffs = mask_coeffs[keep_conf]
    print(f"DEBUG(post_seg): after conf filter -> {xywh.shape[0]} boxes")

    # Convert boxes to xyxy at model input scale (network size)
    boxes_xyxy = ops.xywh2xyxy(xywh)

    # NMS using torchvision, independent of Ultralytics' non_max_suppression signature

    keep_nms = tv_nms(boxes_xyxy, conf, nms_iou_threshold)
    if max_detections is not None:
        keep_nms = keep_nms[:max_detections]

    if keep_nms.numel() == 0:
        print("DEBUG(post_seg): NMS removed all boxes")
        return [], []

    boxes_xyxy = boxes_xyxy[keep_nms]
    conf = conf[keep_nms]
    mask_coeffs = mask_coeffs[keep_nms]
    print(f"DEBUG(post_seg): after NMS -> {boxes_xyxy.shape[0]} boxes")

    # Scale boxes to original image size
    if isinstance(orig_img, list):
        shape = orig_img[0].shape
    else:
        shape = orig_img.shape
    # NOTE: keep float boxes for mask ops; round only for final output serialization
    boxes_xyxy_scaled = ops.scale_boxes(input_hw, boxes_xyxy.clone(), shape)

    # Decode masks using Ultralytics helper
    results = []
    masks = None
    try:
        proto = torch.from_numpy(protos[0])  # (32, Hm, Wm)
        # Prefer Ultralytics' "native" mask path when available (matches .pt inference behavior
        # more closely and tends to produce better-aligned masks).
        #
        # Ultralytics segmentation pipeline:
        #   pred[:, :4] = scale_boxes(imgsz, pred[:, :4], im0.shape)
        #   masks = process_mask_native(proto, pred[:, 6:], pred[:, :4], im0.shape[:2])  (retina_masks)
        #   or masks = process_mask(proto, pred[:, 6:], pred[:, :4], im0.shape[:2], upsample=True)
        #
        # Here, boxes_xyxy_scaled are already in original-image coords.
        if hasattr(ops, "process_mask_native"):
            masks = ops.process_mask_native(
                proto,
                mask_coeffs,
                boxes_xyxy_scaled,
                shape[:2],
            )
        else:
            masks = ops.process_mask(
                proto,
                mask_coeffs,
                boxes_xyxy_scaled,
                shape[:2],        # (H, W) of original image
                upsample=True,
            )

        print(f"DEBUG(post_seg): decoded masks shape {masks.shape}")
    except Exception as e:
        print(f"WARNING(post_seg): mask decoding failed: {e}")
        masks = None

    # Build result list: return tuple format matching postprocess_V8
    # Format: (class_id, x, y, width, height, confidence)
    # Same as postprocess_V8 for compatibility with infer2 workflow
    # Also store masks separately for later retrieval
    masks_list = []
    for i in range(boxes_xyxy_scaled.shape[0]):
        box_xyxy = boxes_xyxy_scaled[i].clone()
        # Convert box to xywh format (same as postprocess_V8)
        box_xywh = ops.xyxy2xywh(box_xyxy.unsqueeze(0))[0]
        
        # Return tuple format: (class_id, x, y, width, height, confidence)
        # class_id is always 0 for single-class segmentation model
        cls = 0
        conf_value = round(float(conf[i]), 2)
        # Round boxes for downstream int() usage in processDefectsAI
        xywh_list = box_xywh.round().tolist()
        
        results.append((int(cls), *xywh_list, conf_value))
        
        # Store mask if available (convert to numpy for serialization)
        if masks is not None and i < len(masks):
            mask_np = masks[i].cpu().numpy()
            masks_list.append(mask_np)
        else:
            masks_list.append(None)

    # Return both detections and masks
    # Store masks in a global dict keyed by a unique identifier
    return results, masks_list
    
what is wrong with my postprocessing ??

**Comment by UltralyticsAssistant** [comment]
👋 Hello @supergamefury, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the detailed context around OpenVINO + segmentation and the “first `predict()` penalty” behavior—this is super helpful for the team to investigate 🔎🙂. Note: this is an automated response, and an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. Ideally, your MRE includes:
-Exact export command you used to produce the `*_openvino_model/` bundle
-A small sample input (or a description of its shape/dtype) and the exact `YOLO(...).predict(...)` call you’re timing
-Your timing method (what you include/exclude: preprocessing, inference, postprocess)
-Complete console logs from the first and second calls

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
Yes—on Intel iGPU it’s common for OpenVINO to look “loaded” quickly but still pay a big one-time cost on the first request (GPU plugin initialization, kernel compilation, memory pool setup), and that cost will repeat if you frequently swap models unless you use OpenVINO’s caching/first-inference mitigation options described in our [OpenVINO latency vs throughput optimization guide](https://docs.ultralytics.com/guides/optimizing-openvino-latency-vs-throughput-modes/).

Ultralytics doesn’t currently expose a “compile/init only” hook separate from running one inference, so the practical approach is still a warmup call; to make it cheaper you can warm up with a dummy frame and a very high `conf` to minimize downstream work:

```python
import numpy as np
from ultralytics import YOLO

m = YOLO("your_model_openvino_model/", task="segment")
_ = m.predict(np.zeros((1024, 1024, 3), dtype=np.uint8), imgsz=1024, device="intel:gpu", conf=0.999, verbose=False)
```

On your custom postprocess, the most common reason it won’t match Ultralytics is a subtle preprocessing/letterbox mismatch (ratio/pad) rather than NMS itself; if you want parity outside Ultralytics you’ll need to replicate Ultralytics’ exact resize/letterbox path and then apply the same box/mask scaling/cropping assumptions, otherwise masks/counts will drift even if the raw tensors look correct.

If you can paste the output of `yolo checks` and confirm whether the 9–20s spike happens with `device="intel:cpu"` too (or only `intel:gpu`), that will help narrow down whether this is GPU kernel compile/caching vs something else in the pipeline.
